base_layer/wallet/src/ffi.rs

```diff
@@ -0,0 +1,511 @@
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+// products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+//! # LibWallet API Definition
+//! This module contains the Rust backend implementations of the functionality that a wallet for the Tari Base Layer
+//! will require. The module contains a number of sub-modules that are implemented as async services. These services are
+//! collected into the main Wallet container struct which manages spinning up all the component services and maintains a
+//! collection of the handles required to interact with those services.
+//! This files contians the API calls that will be exposed to external systems that make use of this module. The API
+//! will be exposed via FFI and will consist of API calls that the FFI client can make into the Wallet module and a set
+//! of Callbacks that the client must implement and provide to the Wallet module to receive asynchronous replies and
+//! updates.
+
+use crate::output_manager_service::service::PendingTransactionOutputs;
+use chrono::NaiveDateTime;
+use std::os::raw::{c_char, c_int, c_uint, c_ulonglong};
+use tari_comms::connection::NetAddress;
+use tari_core::{
+    transaction::{Transaction, UnblindedOutput},
+    types::{PrivateKey, PublicKey},
+};
+
+// use tari_core::transaction_protocol::sender::RawTransactionInfo;
+
+// TODO implement this
+pub struct WalletFfi {}
+
+pub struct IdentityFfi {
+    public_key: PublicKey,
+    secret_key: PrivateKey,
+}
+
+pub struct KeyManagerStateFfi {
+    master_seed: PrivateKey,
+    branch_seed: String,
+    index: c_uint,
+}
+
+pub struct KeyManagerSeedWords {
+    words: Vec<String>,
+}
+
+pub struct NetworkStatusFfi {}
+
+/// Returns a pointer to an initialized Wallet Ffi container. This container needs to be populated with the persistent
+/// data before it can be started.
+/// The generate_identity() function can provide the public_key, private_key data for you
+#[no_mangle]
+pub unsafe extern "C" fn create_wallet(
+    // Local Node Identity data
+    public_key: *const c_char,         // PublicKey
```

Comment:
Agreed, though actually the C type would be the same (byte = char in C) but we will need to include a length as it will not be a null terminated string. Will update the comments to reflect this.